src/sources/transient.rs

```diff
@@ -103,53 +141,99 @@ pub enum TransientSource<T> {
     /// registration, we need to replace the enum variant safely. This requires
     /// having a `None` value in there temporarily while we do the swap.
     ///
     /// If the variant is `None` the value will not change and `replacer` will
-    /// not be called.
+    /// not be called. If the variant is `Replace` then `replacer` will be
+    /// called **on the new source**, which may cause the old source to leak
+    /// registration in the event loop if it has not yet been unregistered.
     ///
     /// The `replacer` function here is expected to be one of the enum variant
     /// constructors eg. `replace(TransientSource::Remove)`.
-    fn replace<F>(&mut self, replacer: F)
+    fn replace_state<F>(&mut self, replacer: F)
     where
         F: FnOnce(T) -> Self,
     {
-        *self = match std::mem::replace(self, TransientSource::None) {
-            TransientSource::Keep(source)
-            | TransientSource::Register(source)
-            | TransientSource::Remove(source)
-            | TransientSource::Disable(source) => replacer(source),
-            TransientSource::None => return,
+        *self = match std::mem::take(self) {
+            Self::Keep(source)
+            | Self::Register(source)
+            | Self::Remove(source)
+            | Self::Disable(source)
+            | Self::Replace { new: source, .. } => replacer(source),
+            Self::None => return,
         };
     }
 }
 
+impl<T> TransientSource<T> {
+    /// Apply a function to the enclosed source, if it exists and is not about
+    /// to be removed.
+    pub fn map<F, U>(&mut self, f: F) -> Option<U>
+    where
+        F: FnOnce(&mut T) -> U,
+    {
+        match &mut self.state {
+            TransientSourceState::Keep(source)
+            | TransientSourceState::Register(source)
+            | TransientSourceState::Disable(source)
+            | TransientSourceState::Replace { new: source, .. } => Some(f(source)),
+            TransientSourceState::Remove(_) | TransientSourceState::None => None,
+        }
+    }
+
+    /// Returns `true` if there is no wrapped event source.
+    pub fn is_none(&self) -> bool {
+        matches!(self.state, TransientSourceState::None)
+    }
+
+    /// Removes the wrapped event source from the event loop and this wrapper.
+    pub fn remove(&mut self) {
+        self.state.replace_state(TransientSourceState::Remove);
+    }
+
+    /// Replace the currently wrapped source with the given one.  No more events
+    /// will be generated from the old source after this point. The old source
+    /// will not be dropped immediately, it will be kept so that it can be
+    /// deregistered.
+    ///
+    /// If this is called from outside of the event loop, you will need to wake
+    /// up the event loop for any changes to take place. If it is called from
+    /// within the event loop, the sources will be registered and unregistered
+    /// as needed after the current `process_events()` iteration.
```

Comment:
... assuming you correctly forward the `PostAction` output, if you are using `TransientSource` as an implementation detail of your own event source.

Reply:
Oops, I put this in the type-level docs but not here. It should be in both places.